### PR TITLE
Clearing redux on logout

### LINKED ISSUE #147: Switching users leads to cache badness
## Issue - Bug

Take the following steps:

1. Login as one user ("user1"), peruse the dashboard
1. Logout
1. Login as a different user ("user2"), ideally with different team memberships
1. Note that user2 lists user1 team memberships. Uh-oh!

Perhaps we should completely invalidate the persisted store if the user id changes?

- **OS:** OSX
- **Browser:** Chrome
- **Node version:** v6.2.0
- **NPM version:** 3.8.9
- **RethinkDB version:** rethinkdb 2.3.4 (CLANG 7.3.0 (clang-703.0.31))
- **Estimated effort:** 2 points ([see CONTRIBUTING.md](https://github.com/ParabolInc/action/blob/master/CONTRIBUTING.md#points-and-sizes))(https://docs.google.com/document/d/1lgTJnJzBt9-4If883iOvthmc0hy-6C6hBF7KNN-lsXU/edit))

### PR DESCRIPTION
**Pull request by jordanh**
Fixes #147

### CODE REVIEW

File: src/universal/routes/logout.js [hunk 1]
```diff
@@ -1,7 +1,16 @@
...
+const reducerWhitelist = [
+  'notifications'
+];
+
+export default (store) => ({
   path: '/logout',
   getComponent: async(location, cb) => {
     const component = await System.import('universal/containers/Logout/LogoutContainer');
+    const newReducer = makeWhitelistedReducer(reducerWhitelist);
+    store.replaceReducer(newReducer);
```
Comment (jordanh) [inline src/universal/routes/logout.js]: Sadly, this doesn't work (yet). Here's what's going on.

1. Navigate to `/me`, then hit gear on UserHub to signout. You'll get to `/me/settings`.
2. Hit Sign Out
3. Router routers to `/logout`

...then `MeSettings.js` re-renders one last time as it unmounts and, guess what?, userSettings reducer is no longer there! Which causes a case of the ickys.

See possible fix in comment above...

File: src/universal/redux/makeReducer.js [hunk 2]
```diff
@@ -6,15 +6,27 @@ import authToken from './authDuck';
...
+const initialReducers = {
   authToken,
   cashay: cashayReducer,
   form: formReducer,
   notifications
 };
+const currentReducers = {...initialReducers};
 
 export default (newReducers) => {
   Object.assign(currentReducers, newReducers);
```
Comment (jordanh) [inline src/universal/redux/makeReducer.js]: Possible fix, I think we may have to keep extending `initialReducers` as well with the initial states of any new reducers we've loaded. What do you think @mattkrick?

File: src/universal/redux/makeReducer.js [hunk 3]
```diff
@@ -6,15 +6,28 @@ import authToken from './authDuck';
...
+const initialReducers = {
   authToken,
   cashay: cashayReducer,
   form: formReducer,
   notifications
 };
+const currentReducers = {...initialReducers};
 
 export default (newReducers) => {
+  Object.assign(initialReducers, newReducers);
```
Comment (jordanh) [inline src/universal/redux/makeReducer.js]: This fixes it...

File: src/universal/routes/logout.js (hunk 1 above)
Comment (jordanh) [inline src/universal/routes/logout.js]: I think [this](https://github.com/ParabolInc/action/pull/157/files#diff-3f48aadcc1b3be42a6f79e920d133c59R18)'ll do it.

**Comment by jordanh** [comment]
Alright @mattkrick, I think this is GTG.

File: src/universal/redux/makeReducer.js (hunk 3 above)
Comment (mattkrick) [inline src/universal/redux/makeReducer.js]: shouldn't that be: `Object.assign(currentReducers, newReducers);`?
If we mutate the initialReducers, we'll only be able to do this trick once

File: src/universal/redux/makeReducer.js [hunk 4]
```diff
@@ -6,15 +6,41 @@ import authToken from './authDuck';
...
+const currentReducers = {...initialReducers};
 
 export default (newReducers) => {
+  /**
+   * Add the initial state of the new reducers to the initialReducers object. If
+   * we later revert the application to the initial state -- say to logout --
+   * we'll need these dynamically loaded initial states present so
+   * mapStateToProps can run one last time before the component unmounts:
+   */
+  Object.assign(initialReducers, newReducers);
```
Comment (jordanh) [inline src/universal/redux/makeReducer.js]: Actually, mutating initialReducers is _exactly_ what we want to do. We need the initial states of any dynamically loaded reducers added here to, or bad things happen in mSTP of dynamically loaded components just before they are unmounted.

If you'd like to see what I mean, try commenting this out and see what happens. I think it'll be fairly obvious...

Procedure:

1. Open console
1. Log in
1. Click gear in hub
1. Log out

Comment (mattkrick) [inline src/universal/redux/makeReducer.js]: i think i see what you're saying, but my brain is fried right now. gonna grab some grub & have another look.

Comment (jordanh) [inline src/universal/redux/makeReducer.js]: When you have a look, let me know, I think this is ready now.

Comment (mattkrick) [inline src/universal/redux/makeReducer.js]: Still gotta dig into this. There's got to be an eloquent way to unsubscribe from all of those listeners before calling `dispatch`. Will PR tomorrow

Comment (jordanh) [inline src/universal/redux/makeReducer.js]: Ah, of course. Silly/stupid way would be to aggregate all of the unsubscribe functions into an array stored in a duck. Then, when logout happens, just call each unsubscribe handler...

Comment (mattkrick) [inline src/universal/redux/makeReducer.js]: well the `connect` hoc should unsubscribe when the component gets unmounted, which is what had me confused. I'm _almost_ out of the weeds on cashay & have action working with no errors, will wrap up shortly

**Comment by mattkrick** [comment]
In my tests, this still keeps the cashay data alive. That ain't good, since domain state data could be sensitive. Are you getting the same results?

**Comment by jordanh** [comment]
Alas @mattkrick you were right. I didn't test the original fix deeply enough. Reading Dan Abramov's answer to [this Stackoverflow question](http://stackoverflow.com/questions/35622588/how-to-reset-the-state-of-a-redux-store) ended up being a real forehead-slapper.

Now we're simply wrapping our application reducer in a new rootReducer that can return a reset application state. I believe this is the sledgehammer we've been looking for.

I also added some unit tests, cause, we're good like that.

File: src/universal/redux/rootDuck.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+
+export const RESET = '@@root/RESET';
+
+export default (appReducer) => (state, action) => {
+  if (action && action.type === RESET) {
+    const {whitelist} = action.payload;
+    const newState = {};
+    whitelist.forEach((keeper) => {
+      if (keeper in state) {
+        newState[keeper] = state[keeper];
```
Comment (jordanh) [inline src/universal/redux/rootDuck.js]: Just making sure shallow is enough here: https://twitter.com/dan_abramov/status/688087202312491008

Comment (mattkrick) [inline src/universal/redux/rootDuck.js]: yeah, this is fine since we'll never mutate `state[keeper]`.

**Comment by jordanh** [comment]
Ok, now I really think this is GTG.

@mattkrick we all clear?

**Pull request merged**